Annotation processor 'io.vertx.codegen.CodeGenProcessor' not found & the module-name conflict

Hi! While compiling my Vertx project using `vertx-service-proxy`, I found I get an confusing error: `Annotation processor 'io.vertx.codegen.CodeGenProcessor' not found`

Here is the pom.xml (Or you can get the whole project here: https://github.com/CasterKKK/CasterMovieVertx if you want):
```
<dependencies>
    ...
    <dependency>
        <groupId>io.vertx</groupId>
        <artifactId>vertx-service-proxy</artifactId>
        <classifier>processor</classifier>
    </dependency>
    ...
</dependencies>

...
<plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>${maven-compiler.version}</version>
        <configuration>
            <source>${java.version}</source>
            <target>${java.version}</target>
            <useIncrementalCompilation>false</useIncrementalCompilation>
            <encoding>${project.build.sourceEncoding}</encoding>
            <generatedSourcesDirectory>src/main/generated</generatedSourcesDirectory>
            <generatedTestSourcesDirectory>src/test/generated</generatedTestSourcesDirectory>
         </configuration>
           <executions>
                <execution>
                    <id>default-compile</id>
                        <configuration>
                            <annotationProcessors>
                                <annotationProcessor>io.vertx.codegen.CodeGenProcessor</annotationProcessor>
                            </annotationProcessors>
                            <generatedSourcesDirectory>src/main/generated</generatedSourcesDirectory>
                            <compilerArgs>
                                <arg>-Acodegen.output=${project.basedir}/src/main</arg>
                            </compilerArgs>
                        </configuration>
                    </execution>
                    <execution>
                        <id>default-testCompile</id>
                        <configuration>
                            <annotationProcessors>
                                <annotationProcessor>io.vertx.codegen.CodeGenProcessor</annotationProcessor>
                            </annotationProcessors>
                           <generatedTestSourcesDirectory>src/test/generated</generatedTestSourcesDirectory>
                       <compilerArgs>
                          <arg>-Acodegen.output=${project.basedir}/src/test</arg>
                    </compilerArgs>
                </configuration>
             </execution>
         </executions>
</plugin>
```

Show one of the Service:
```
@VertxGen
@ProxyGen
public interface UserService<T extends User> {

    @Fluent
    UserService register(T element, Handler<AsyncResult<T>> resultHandler) throws IOException;

    @Fluent
    UserService validate(UUID id, Handler<AsyncResult<T>> resultHandler) throws IOException;

    @Fluent
    UserService cancel(UUID id, Handler<AsyncResult<T>> resultHandler);

    @Fluent
    UserService update(UUID id, T element, Handler<AsyncResult<T>> resultHandler);

    @Fluent
    UserService getByJwt(String jwt, Handler<AsyncResult<T>> resultHandler);

    @Fluent
    UserService findById(UUID id, Handler<AsyncResult<T>> resultHandler);

    @Fluent
    UserService findByEmail(String email, Handler<AsyncResult<T>> resultHandler);

    @Fluent
    UserService findAllByState(State state, Handler<AsyncResult<List<T>>> resultHandler);

    @Fluent
    UserService findAllByCreateTimeBetween(Instant from, Instant to, Handler<AsyncResult<List<T>>> resultHandler);

    @Fluent
    UserService findAllByCancelledTimeBetween(Instant from, Instant to, Handler<AsyncResult<List<T>>> resultHandler);
}
```

And when using `mvn package -X -f pom.xml`, get an error:
```
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.7.0:compile (default-compile) on project user-module: Compilation failure
Annotation processor 'io.vertx.codegen.CodeGenProcessor' not found

	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
	at org.codehaus.classworlds.Launcher.main(Launcher.java:47)
Caused by: org.apache.maven.plugin.compiler.CompilationFailureException: Compilation failure
Annotation processor 'io.vertx.codegen.CodeGenProcessor' not found

	at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:1161)
	at org.apache.maven.plugin.compiler.CompilerMojo.execute(CompilerMojo.java:168)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	... 21 more
[ERROR] 
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :user-module
```

---

What's more, I found when I use
```
        <dependency>
            <groupId>io.vertx</groupId>
            <artifactId>vertx-service-proxy</artifactId>
            <classifier>processor</classifier>
        </dependency>
```
and 
```
<dependency>
    <groupId>io.vertx</groupId>
    <artifactId>vertx-codegen</artifactId>
</dependency>
```
together, my `module-info.java` will get a `Module xxx reads package 'io.vertx.codegen' from both 'vertx.codegen' and 'vertx.codegen'` & `Ambiguous module reference: vertx.codege` error... What that mean?

--- a/README.md
+++ b/README.md
@@ -87,7 +87,7 @@ You can configure the `CodeGenProcessor` as any Java annotation processor, here 
     <!-- Configure the execution of the compiler to execute the codegen processor -->
     <plugin>
       <artifactId>maven-compiler-plugin</artifactId>
-      <version>3.1</version>
+      <version>3.8.1</version>
       <configuration>
         <source>1.8</source>
         <target>1.8</target>
@@ -102,6 +102,15 @@ You can configure the `CodeGenProcessor` as any Java annotation processor, here 
             <annotationProcessors>
               <annotationProcessor>io.vertx.codegen.CodeGenProcessor</annotationProcessor>
             </annotationProcessors>
+             <!-- It is new option since v3.5 to instruct compiler detect annotation processors classpath -->
+            <annotationProcessorPaths>
+              <path>
+                <groupId>io.vertx</groupId>
+                <artifactId>vertx-codegen</artifactId>
+                <version>${vertx.version}</version>
+              </path>
+              <!-- ... more path such as vertx-service-proxy/vertx-rx-java2 depends on what you want to generate ... -->
+            </annotationProcessorPaths>
             <compilerArgs>
               <arg>-Acodegen.output=${project.basedir}/src/main</arg>
             </compilerArgs>
@@ -113,13 +122,23 @@ You can configure the `CodeGenProcessor` as any Java annotation processor, here 
 </pluginManagement>
 ~~~~
 
-And here is a configuration example for Gradle:
+And here is a configuration example for Gradle (since `Gradle 5.0`):
+
+Gradle Groovy
 
 ```gradle
-task annotationProcessing(type: JavaCompile, group: 'build') { // codegen
+dependencies {
+    compileOnly("io.vertx:vertx-codegen:4.0.2")
+    // more vertx-service-proxy/vertx-rx-java2
+    // compileOnly("io.vertx:vertx-rx-java2:4.0.2")
+    annotationProcessor("io.vertx:vertx-codegen:4.0.2")
+}
+
+task annotationProcessing(type: JavaCompile, group: 'other') { // codegen
   source = sourceSets.main.java
-  classpath = configurations.compile + configurations.compileOnly
-  destinationDir = project.file('src/main/generated')
+  classpath = configurations.compile
+  destinationDir = project.file('${project.buildDir}/generated/main/java')
+  options.annotationProcessorPath = configurations.compileClasspath
   options.compilerArgs = [
     "-proc:only",
     "-processor", "io.vertx.codegen.CodeGenProcessor",
@@ -128,25 +147,58 @@ task annotationProcessing(type: JavaCompile, group: 'build') { // codegen
 }
 
 compileJava {
-  targetCompatibility = 1.8
-  sourceCompatibility = 1.8
-
   dependsOn annotationProcessing
 }
 
 sourceSets {
   main {
     java {
-      srcDirs += 'src/main/generated'
+      srcDirs += '${project.buildDir}/generated/main/java'
     }
   }
+}
+```
+
+Gradle Kotlin
+
+```gradle
+dependencies {
+    compileOnly("io.vertx:vertx-codegen:4.0.2")
+    // more vertx-service-proxy/vertx-rx-java2
+    // compileOnly("io.vertx:vertx-rx-java2:4.0.2")
+    annotationProcessor("io.vertx:vertx-codegen:4.0.2")
+}
+
+tasks.register<JavaCompile>("annotationProcessing") {
+    group = "other"
+    source = sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME).java
+    destinationDir = project.file("${project.buildDir}/generated/main/java")
+    classpath = configurations.compileClasspath.get()
+    options.annotationProcessorPath = configurations.compileClasspath.get()
+    options.compilerArgs = listOf(
+        "-proc:only",
+        "-processor", "io.vertx.codegen.CodeGenProcessor",
+        "-Acodegen.output=${project.projectDir}/src/main"
+    )
+}
+
+tasks.compileJava {
+    dependsOn(tasks.named("annotationProcessing"))
+}
+
+sourceSets {
+    main {
+        java {
+            srcDirs(project.file("${project.buildDir}/generated/main/java"))
+        }
+    }
 }
 ```
 
 Besides you can use the `processor` classified dependency that declares the annotation processor as a
 `META-INF/services/javax.annotation.processing.Processor`, if you do so, code generation happens automatically:
 
-```
+```xml
 <dependency>
   <groupid>io.vertx</groupId>
   <artifactId>vertx-codegen</artifactId>

